### ISSUE TITLE
Parsing the CoBiz data file

### ISSUE THREAD

**Issue by jcasimir** [turingschool/monocle#2]
We need to write a parser that can take that mess of text in `data/` and turn it into a clean CSV or JSON.

**Comment by indiesquidge** [comment]
@jcasimir, what is the expected result? Are there particular fields we want to concatenate, leave out, or alter (e.g. whether or not the address should be a single value)? What should the JSON keys be? Etc.

I've taken a stab at what I think the parsed data should look like. Does this seem reasonable?

*Before*:
```txt
ADVANCED THIN FILMS LLC
5733 Central Ave.
Boulder, CO 80301
303-815-1545
www.atf-ppc.com
Company headquarters: Boulder
Products/Services: Precision optical
components and ion beam sputtered coatings
for applications in scientific research, defense,
aerospace, telecommunications, laser and
semiconductor manufacturing.
Person in charge: Robert Beeson, general
manager
```

*After*:
```json
{
  "name": "ADVANCED THIN FILMS LLC",
  "address": "5733 Central Ave. Boulder, CO 80301",
  "phoneNumber": "303-815-1545",
  "website": "www.atf-ppc.com",
  "companyHeadquarters": "Boulder",
  "productsAndServices": "Precision optical components and ion beam sputtered coatings for applications in scientific research, defense, aerospace, telecommunications, laser and semiconductor manufacturing.",
  "personInCharge": "Robert Beeson, general manager"
}
```

**Comment by dglunz** [comment]
Ayy this looks familiar 🙃 We went down this exact scraping path for finding local b2b leads at work. 

For cold sales tho, we didn't really give a shit about anything besides `company name`, `size`, and a `point of contact`. So we used https://clearbit.com/docs#prospector-api to try and find the right person to contact. 

@jcasimir not sure what your end goal with this set of data is but it feels like most of it is junk except maybe `name` + `description`. Are we just trying to get a base of companies to avoid duplication when students go to add their experiences applying?

Parsed in a mostly @indiesquidge format: https://gist.github.com/dglunz/8b8d12aa9b5eadc50fb396f4fa7bea3e

**Comment by stevekinney** [comment]
@dglunz 👍 Nicely done—that said, [it looks like it's slightly off in some edge cases](https://gist.github.com/dglunz/8b8d12aa9b5eadc50fb396f4fa7bea3e#file-parsed-json-L27-L31). I suspect extraneous commas are to blame.

**Comment by calaway** [comment]
@dglunz, your parser looks like it's doing an awesome job on this dataset. I'm going to be working on this for the next two weeks as my Mod 3 final project, so I'd love to build off it, although I have never before worked with coffeescript.

Have you put any thought into how to address the edge cases like the one Steve pointed out above? If so, I'd love to hear it. 

Also, if you're in the Denver area and might be available to pair for an hour or so this week to help me build off of your work that would be really awesome. If that's not possible though, I understand.

**Comment by calaway** [comment]
@indiesquidge, did you write a parser for the one you did? If so, is it in Ruby, by chance? If you have anything you can share with me that I might be able to build off of I would greatly appreciate it.

**Comment by dglunz** [comment]
@calaway yeah sorry about the coffeescript, I figured it was close enough to ruby style 😄  The regex is probably the only worthwhile part of that code anyhow. 

I'd probably take that chunk of data and see if tweaking the regex could fix it. I'm guessing it has to do with titles that overflow to the next line. 

Happy to meet up some morning and help. Slack me and we can figure out a time.

**Comment by calaway** [comment]
I just posted this in the #monocle slack channel:

If anyone wants to help on the data parser, that’s where we could really use it at this point, since I don’t expect to be able to get to that problem and all my work will really be useless without it.

After working with the data I believe what we really need is to sort it into some sort of delimited form, such as CSV. The reason for this is it will be much easier if nil values are still represented, which isn’t the case with YAML or JSON. The way I accomplished this just to make some it work for the text fixtures is to make my own ad hoc format wherein the different company attributes are delimited, or separated, by line breaks. In this each ten lines represents a company, so on a mod 10 basis each line represents:
1. company name
2. street address
3. city, state, zip
4. phone
5. website
6. headquarters
7. products/services
8. person in charge
9. blank
10. blank
Check out this one for example: https://github.com/calaway/monocle/blob/master/data/test_fixtures_data/agricultural_technology.txt

I think I would like to see a service module or class for it (i.e. `./app/services/data_parser.rb`) and from that, if it’s fully automated, we can write a rake task to execute it.
